kaitai/kaitaistream.cpp

```diff
@@ -1197,6 +1197,112 @@ std::string kaitai::kstream::bytes_to_str(const std::string src, int codepage) {
     return utf8;
 }
 
+#elif defined(KS_STR_ENCODING_ICU)
+#include <unicode/ucnv.h>
+#include <unicode/ustring.h>
+
+std::string kaitai::kstream::bytes_to_str(const std::string src, const char *src_enc) {
+    UErrorCode err = U_ZERO_ERROR;
+
+    // Open the source converter
+    UConverter* conv = ucnv_open(src_enc, &err);
+    if (U_FAILURE(err)) {
+        if (err == U_FILE_ACCESS_ERROR) {
+            throw unknown_encoding(src_enc);
+        }
+        throw bytes_to_str_error(u_errorName(err));
+    }
+
+    // Open UTF-8 converter
+    UConverter* utf8Conv = ucnv_open("UTF-8", &err);
+    if (U_FAILURE(err)) {
+        ucnv_close(conv);
+        throw bytes_to_str_error(u_errorName(err));
+    }
+
+    // Configure source converter to stop on illegal sequences
+    err = U_ZERO_ERROR;
+    ucnv_setToUCallBack(
+        conv,
+        UCNV_TO_U_CALLBACK_STOP,
+        nullptr,
+        nullptr,
+        nullptr,
+        &err);
+    if (U_FAILURE(err)) {
+        ucnv_close(conv);
+        ucnv_close(utf8Conv);
+        throw illegal_seq_in_encoding(u_errorName(err));
+    }
+
+    // Allocate buffer for UTF-16 intermediate representation
+    const int32_t uniStrCapacity = UCNV_GET_MAX_BYTES_FOR_STRING(src.length(), ucnv_getMaxCharSize(conv));
+    UChar* uniStr = new UChar[uniStrCapacity];
+
+    // Convert from source encoding to UTF-16
+    err = U_ZERO_ERROR;
+    int32_t uniLength = ucnv_toUChars(
+        conv,
+        uniStr,
+        uniStrCapacity,
+        src.c_str(),
+        src.length(),
+        &err);
+    if (U_FAILURE(err)) {
+        delete[] uniStr;
+        ucnv_close(conv);
+        ucnv_close(utf8Conv);
+        throw illegal_seq_in_encoding(u_errorName(err));
```

Comment:
This is the only place where `illegal_seq_in_encoding` is thrown when running our test suite, and the only place where it actually makes sense to throw it. It makes no sense when handling potential errors from `ucnv_setToUCallBack`, `ucnv_setFromUCallBack` or `ucnv_fromUChars` - if these fail, it doesn't indicate an illegal sequence in the input string (instead, it would indicate a bug in our `bytes_to_str` implementation, as none of these should normally fail regardless of the user input).

Therefore, `throw illegal_seq_in_encoding` should only remain here and `throw bytes_to_str_error` should be used everywhere else.